File: Leetcode/Java/2. Add Two Numbers/solution.java
// https://leetcode.com/problems/add-two-numbers/description/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {    // Full adder
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode root = new ListNode(-1);
        ListNode cur = root;

        while (l1 != null || l2 != null || carry != 0){
            int total = carry;

            if (l1 != null){
                total += l1.val;
                l1 =l1.next;
            }

            if (l2 != null){
                total += l2.val;
                l2 = l2.next;
            }

            carry = (int)(total / 10);
            cur.next = new ListNode(total % 10);
            cur = cur.next;
        }

        return root.next;
    }
}

File: Leetcode/Typescript/2. Add Two Numbers/solution.ts
// https://leetcode.com/problems/add-two-numbers/description/

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(
  l1: ListNode | null,
  l2: ListNode | null
): ListNode | null {
  const root = new ListNode();
  let cur = root,
    carry = 0;

  while (l1 || l2 || carry) {
    let total = carry;

    if (l1) {
      total += l1.val;
      l1 = l1.next;
    }
    if (l2) {
      total += l2.val;
      l2 = l2.next;
    }

    cur.next = new ListNode(total % 10);
    cur = cur.next;
    carry = total >= 10 ? 1 : 0;
  }

  return root.next;
}
